--- src/arc4random.c.orig	2014-03-23 01:17:43.124832361 +0900
+++ src/arc4random.c	2014-03-23 01:19:11.414128014 +0900
@@ -31,7 +31,7 @@
 #include <sys/param.h>
 #include <sys/time.h>
 #include <sys/sysctl.h>
-#include "thread_private.h"
+#include <stdio.h>
 
 #define KEYSTREAM_ONLY
 #include "chacha_private.h"
@@ -67,16 +67,33 @@
 static void
 _rs_stir(void)
 {
-	int     mib[2];
 	size_t	len;
 	u_char rnd[KEYSZ + IVSZ];
 
+#ifdef KERN_ARND
+	int     mib[2];
 	mib[0] = CTL_KERN;
 	mib[1] = KERN_ARND;
 
 	len = sizeof(rnd);
 	sysctl(mib, 2, rnd, &len, NULL, 0);
 
+#else
+        int fd;
+        size_t r;
+        fd = open("/dev/urandom", 0);
+        if (fd == -1) {
+                fprintf(stderr, "can't open random device\n");
+                abort();
+        }
+
+        r = read(fd, rnd, len);
+        if (r == -1) {
+                fprintf(stderr, "can't read random device\n");
+                abort();
+        }
+        close(fd);
+#endif
 	if (!rs_initialized) {
 		rs_initialized = 1;
 		_rs_init(rnd, sizeof(rnd));
